shared/route-tree/index.js

```diff
@@ -0,0 +1,229 @@
+
+type LeafTagsParams = {
+  modal: boolean,
+  underStatusBar: boolean,
+}
+
+export const LeafTags: (spec?: LeafTagsParams) => LeafTagsParams & I.Record<LeafTagsParams> = I.Record({
+  modal: false,
+  underStatusBar: false,
+})
+
+const _RouteDefNode = I.Record({
+  defaultSelected: null,
+  component: null,
+  containerComponent: null,
+  tags: LeafTags(),
+  initialState: I.Map(),
+  children: I.Map(),
+})
+
+type RouteDefParams<P> = {
+  defaultSelected?: string,
+  tags?: LeafTags,
+  initialState?: {},
+  children: {},
+} & (
+  // This lengthy type definition was necessary to get all of our component permutations to type check.
+  { component?: Component<any, P, any> | $Supertype<Component<any, P, any>> | Class<ConnectedComponent<P, any, any, any>> | Class<TypedConnectedComponent<P>> }
+  | { containerComponent: Component<any, P, any> }
+)
+
+export class RouteDefNode extends _RouteDefNode {
+  constructor ({defaultSelected, component, containerComponent, tags, initialState, children}: RouteDefParams<*>) {
+    super({
+      defaultSelected: defaultSelected || null,
+      component,
+      containerComponent,
+      tags: LeafTags(tags),
+      initialState: I.Map(initialState),
+      props: I.Map(),
+      state: I.Map(),
+      children: I.Seq(children)
+        .map(params => params instanceof RouteDefNode || typeof params === 'function' ? params : new RouteDefNode(params))
+        .toMap(),
+    })
+  }
+
+  getChild (name: string): ?RouteDefNode {
+    const childDef = this.children.get(name)
+    if (!childDef) {
+      return
+    }
+    if (typeof childDef === 'function') {
+      return childDef()
+    }
+    return childDef
+  }
+}
+
+type RouteStateParams = {
+  selected: string | null,
+  props?: I.Map<string, any>,
+  state?: I.Map<string, any>,
+}
+
+const _RouteStateNode = I.Record({
+  selected: null,
+  props: I.Map(),
+  state: I.Map(),
+  children: I.Map(),
+})
+
+export class RouteStateNode extends _RouteStateNode {
+  constructor (data: RouteStateParams) {  // eslint-disable-line no-useless-constructor
+    super(data)
+  }
+
+  getChild (name: string): RouteStateNode {
+    return this.children.get(name)
+  }
+
+  updateChild (name: string, op: (node: RouteStateNode) => ?RouteStateNode): RouteStateNode {
+    return this.updateIn(['children', name], op)
+  }
+}
+
+export class InvalidRouteError extends Error {}
+
+// Explicit list of iterable types to accept. We don't want to allow strings
+// since navigateTo('foo') instead of navigateTo(['foo']) is an easy mistake to
+// make.
+type PathIterable<X> = I.IndexedSeq<X> | I.List<X> | Array<X>
+export type Path = PathIterable<string>
+export type PropsPath = PathIterable<string | {selected: string | null}>
+type PathParam = [] | Path | PropsPath  // Flow doesn't accept Path as a subtype of PropsPath, so be explicit here.
+type PathSetSpec = I.IndexedIterable<{type: 'next' | 'navigate', next: string | null, props?: {}}>
```

Comment:
Ah this is tricky, I was going to try to document it but felt like it would balloon into several paragraphs.

`navigate` means update the `selected` value, while `next` doesn't. In either case, you traverse to the value of `next`. This is how `parentPath` descends to a desired path *without* navigating to it. How about we rename `type: 'next'` to `type: 'traverse'`? Still not great, but I am not coming up with a better name at the moment. :)